File: sdk/python/src/attestplane/adapters/langsmith.py
# SPDX-FileCopyrightText: 2026 The Attestplane Authors
# SPDX-License-Identifier: Apache-2.0
"""LangSmith → Attestplane evidence-event adapter.

Implements ticket #31 of the competitive_positioning_upgrade_plan_20260517.

A LangSmith ``Run`` describes one node of an agent / chain / LLM /
tool execution; this adapter translates one ``Run`` into one
:class:`~attestplane.types.EventDraft` per the v1 evidence taxonomy
(``tool_call_event`` family).

The adapter is **pure** per the
:class:`~attestplane.adapters.GenericRuntimeAdapter` contract: no
I/O, no clock reads, no LangSmith API calls. Callers feed in already-
fetched ``Run`` records (production callers typically pipe
LangSmith's ``Client.list_runs()`` output through this adapter; tests
use the :class:`LangSmithRun` dataclass directly).

Trust boundary (ADR-0004 § 4):

- This adapter lives in the Attestplane OSS substrate tree because
  LangSmith's Run shape is public documentation, not LangChain
  proprietary code. The adapter consumes a shape, not a runtime.
- LangSmith / LangChain do NOT endorse this adapter; it is built
  from publicly documented trace structures per
  ``docs/policy/forbidden_claims.md § G``.

Redaction (mandatory per ADR-0008 § Boundary anti-requirements):

- ``Run.inputs`` and ``Run.outputs`` are NEVER copied into the
  evidence event payload as raw values. They are hashed
  (SHA-256 hex) and the hash is placed in ``arguments_hash`` /
  ``result_hash``.
- Tool / chain names go into ``payload.tool_name``.
- Errors go into ``payload.error_code`` (the LangSmith error string,
  truncated to 200 chars to avoid accidental PII leak from
  exception traces).
- ``parent_run_id`` and ``trace_id`` map to ``session_id``.

Run type → event_type mapping:

| LangSmith ``run_type``       | Attestplane ``event_type`` |
|------------------------------|----------------------------|
| ``tool``                     | ``tool_call_event``        |
| ``llm``                      | ``tool_call_event``        |
| ``chain``                    | ``tool_call_event``        |
| ``retriever``                | ``tool_call_event``        |
| ``prompt``                   | ``tool_call_event``        |
| ``parser``                   | ``tool_call_event``        |
| ``embedding``                | ``tool_call_event``        |
| anything else                | ``tool_call_event`` (with ``payload.kind="unknown"``) |

All run types map to ``tool_call_event`` in v1; future taxonomy
revisions may split (e.g., separate ``llm_call_event`` predicate).
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any

from attestplane.adapters.base import AdapterTranslationError, GenericRuntimeAdapter
from attestplane.event_types import TOOL_CALL_EVENT
from attestplane.types import EventDraft, SubjectRef

_KNOWN_RUN_TYPES: frozenset[str] = frozenset({
    "tool", "llm", "chain", "retriever",
    "prompt", "parser", "embedding",
})


@dataclass(frozen=True, slots=True)
class LangSmithRun:
    """Subset of LangSmith's ``Run`` schema that this adapter consumes.

    Fields follow the public LangSmith API documentation
    (https://docs.smith.langchain.com/). Fields not listed here are
    accepted in the input dict by :meth:`LangSmithAdapter.from_dict`
    but ignored — the adapter is conservative about what it reads.
    """

    id: str
    name: str
    run_type: str
    start_time: datetime
    end_time: datetime | None = None
    inputs: dict[str, Any] = field(default_factory=dict)
    outputs: dict[str, Any] | None = None
    error: str | None = None
    trace_id: str | None = None
    parent_run_id: str | None = None
    status: str | None = None
    tags: tuple[str, ...] = ()
    metadata: dict[str, Any] = field(default_factory=dict)
    end_user_id: str | None = None
    """LangSmith metadata.user_id — wrapped in SubjectRef if present."""


def _hash_json(value: Any) -> str:
    """Deterministic SHA-256 hex of a JSON-serializable value.

    Uses sorted keys + compact separators so equivalent dicts hash
    identically regardless of key insertion order.
    """
    encoded = json.dumps(value, sort_keys=True, separators=(",", ":"),
                         default=str, ensure_ascii=False).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


def _truncate(text: str, n: int = 200) -> str:
    if len(text) <= n:
        return text
    return text[:n - 3] + "..."


class LangSmithAdapter(GenericRuntimeAdapter[LangSmithRun]):
    """Translates one :class:`LangSmithRun` into one ``EventDraft``."""

    runtime_name = "langsmith"
    schema_version = 1

    def translate(self, runtime_event: LangSmithRun) -> EventDraft:
        if not isinstance(runtime_event, LangSmithRun):
            raise AdapterTranslationError(
                f"expected LangSmithRun, got {type(runtime_event).__name__}"
            )
        run = runtime_event

        # Determine result_status from status / error.
        if run.error:
            result_status = "ERROR"
        elif run.end_time is None and run.status in (None, "in_progress"):
            # Run still running at trace-export time — emit nothing
            # confusing; mark as OK (the runtime will emit the
            # terminal event when it completes).
            result_status = "OK"
        else:
            result_status = "OK"

        kind = run.run_type if run.run_type in _KNOWN_RUN_TYPES else "unknown"

        payload: dict[str, Any] = {
            "kind": kind,
            "tool_name": f"langsmith.{kind}.{run.name}",
            "tool_call_id": run.id,
            "arguments_hash": _hash_json(run.inputs),
            "result_status": result_status,
        }
        if run.outputs is not None:
            payload["result_hash"] = _hash_json(run.outputs)
        if run.end_time is not None:
            duration_ms = int(
                (run.end_time - run.start_time).total_seconds() * 1000
            )
            payload["latency_ms"] = duration_ms
        if run.error:
            payload["error_code"] = _truncate(run.error)
        if run.tags:
            payload["tags"] = list(run.tags)

        subject_ref: SubjectRef | None = None
        if run.end_user_id:
            # Pseudonymize the LangSmith user_id field; sha256_salted
            # is the substrate's strongest scheme. Adapters that have
            # a stable salt pre-hash before passing; we use the raw
            # value here and tag scheme=opaque (the upstream is
            # already an opaque ID, not raw PII).
            subject_ref = SubjectRef(scheme="opaque", value=run.end_user_id)

        # session_id: prefer trace_id (LangSmith trace), fall back to
        # the run's own id (single-event "session").
        session_id = run.trace_id or run.id

        return EventDraft(
            event_type=TOOL_CALL_EVENT,
            actor=f"langsmith://{run.run_type}/{run.name}",
            payload=payload,
            subject_ref=subject_ref,
            session_id=session_id,
            reference_db_ref=run.parent_run_id,
        )

    @staticmethod
    def from_dict(raw: dict[str, Any]) -> LangSmithRun:
        """Build a :class:`LangSmithRun` from a LangSmith API dict.

        Defensive: accepts the public Run schema, ignores fields we
        don't model. Raises :class:`AdapterTranslationError` on
        missing required fields or malformed datetimes.
        """
        required = ("id", "name", "run_type", "start_time")
        missing = [k for k in required if k not in raw]
        if missing:
            raise AdapterTranslationError(
                f"LangSmith run dict missing required fields: {sorted(missing)}"
            )

        def parse_dt(value: Any) -> datetime:
            if isinstance(value, datetime):
                return value
            if isinstance(value, str):
                # LangSmith uses ISO 8601; some payloads use 'Z' for UTC.
                normalized = value.replace("Z", "+00:00") if value.endswith("Z") else value
                try:
                    return datetime.fromisoformat(normalized)
                except ValueError as exc:
                    raise AdapterTranslationError(
                        f"unparsable datetime {value!r}: {exc}"
                    ) from exc
            raise AdapterTranslationError(
                f"datetime field has type {type(value).__name__}, "
                f"expected datetime or ISO string"
            )

        metadata = raw.get("metadata") or {}
        if not isinstance(metadata, dict):
            raise AdapterTranslationError("metadata must be a dict")

        tags_raw = raw.get("tags") or ()
        if not isinstance(tags_raw, (list, tuple)):
            raise AdapterTranslationError("tags must be a list or tuple")

        return LangSmithRun(
            id=str(raw["id"]),
            name=str(raw["name"]),
            run_type=str(raw["run_type"]),
            start_time=parse_dt(raw["start_time"]),
            end_time=parse_dt(raw["end_time"]) if raw.get("end_time") else None,
            inputs=raw.get("inputs") or {},
            outputs=raw.get("outputs"),
            error=raw.get("error"),
            trace_id=raw.get("trace_id"),
            parent_run_id=raw.get("parent_run_id"),
            status=raw.get("status"),
            tags=tuple(str(t) for t in tags_raw),
            metadata=metadata,
            end_user_id=metadata.get("user_id") if isinstance(metadata, dict) else None,
        )


__all__ = [
    "LangSmithAdapter",
    "LangSmithRun",
]

File: sdk/typescript/src/adapters/langsmith.ts
// SPDX-FileCopyrightText: 2026 The Attestplane Authors
// SPDX-License-Identifier: Apache-2.0
/**
 * LangSmith → Attestplane evidence-event adapter (TypeScript port of
 * `sdk/python/src/attestplane/adapters/langsmith.py`).
 *
 * Byte-identical translation semantics with the Python adapter:
 * given the same LangSmith Run, both adapters produce the same
 * EventDraft fields (event_type, actor, payload, subject_ref,
 * session_id, reference_db_ref).
 *
 * Pure function per GenericRuntimeAdapter contract: no I/O, no
 * LangSmith API calls.
 *
 * Trust boundary (ADR-0004 § 4): adapter consumes the public
 * LangSmith Run shape, not LangChain proprietary code. LangChain
 * does not endorse this adapter; see docs/policy/forbidden_claims.md
 * § G.
 *
 * Redaction (ADR-0008 § Boundary anti-requirements):
 * - Raw inputs / outputs are NEVER copied to payload — only SHA-256
 *   hashes go in `arguments_hash` / `result_hash`.
 * - Error strings are truncated to 200 chars to limit PII leakage.
 * - end_user_id (from LangSmith metadata.user_id) is wrapped in
 *   SubjectRef(scheme="opaque").
 *
 * Run type → event_type mapping is identical to the Python adapter:
 * all run_types map to `tool_call_event` in v1; payload.kind carries
 * the LangSmith run_type for downstream filtering.
 */

import { createHash } from 'node:crypto';

import { AdapterTranslationError, GenericRuntimeAdapter } from '../adapters.js';
import { TOOL_CALL_EVENT } from '../event_types.js';
import { makeEventDraft, makeSubjectRef, type EventDraft } from '../types.js';

const KNOWN_RUN_TYPES = new Set([
  'tool', 'llm', 'chain', 'retriever',
  'prompt', 'parser', 'embedding',
]);

export interface LangSmithRun {
  readonly id: string;
  readonly name: string;
  readonly run_type: string;
  readonly start_time: Date;
  readonly end_time?: Date | null;
  readonly inputs?: Record<string, unknown>;
  readonly outputs?: Record<string, unknown> | null;
  readonly error?: string | null;
  readonly trace_id?: string | null;
  readonly parent_run_id?: string | null;
  readonly status?: string | null;
  readonly tags?: readonly string[];
  readonly metadata?: Record<string, unknown>;
  readonly end_user_id?: string | null;
}

function _canonicalReplacer(): (this: unknown, key: string, value: unknown) => unknown {
  // sort_keys + default=str equivalent (matches Python's json.dumps
  // with sort_keys=True, separators=(",",":"), default=str).
  return function (this: unknown, _key: string, value: unknown): unknown {
    if (value === undefined) return undefined;
    if (value instanceof Date) return value.toISOString();
    if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
      const sorted: Record<string, unknown> = {};
      const obj = value as Record<string, unknown>;
      for (const k of Object.keys(obj).sort()) {
        sorted[k] = obj[k];
      }
      return sorted;
    }
    return value;
  };
}

function _hashJson(value: unknown): string {
  // Match Python: json.dumps(value, sort_keys=True, separators=(",",":"), default=str)
  // JS JSON.stringify by default uses no separators; we need to mimic compact form.
  // The canonical_replacer above sorts keys; we then re-emit with the no-space
  // separator form by passing the result through a second stringify.
  const sorted = JSON.parse(JSON.stringify(value, _canonicalReplacer()));
  const compact = JSON.stringify(sorted);
  return createHash('sha256').update(compact, 'utf-8').digest('hex');
}

function _truncate(text: string, n = 200): string {
  if (text.length <= n) return text;
  return text.slice(0, n - 3) + '...';
}

export class LangSmithAdapter extends GenericRuntimeAdapter<LangSmithRun> {
  readonly runtime_name = 'langsmith';
  readonly schema_version = 1;

  translate(runtime_event: LangSmithRun): EventDraft {
    if (
      typeof runtime_event !== 'object'
      || runtime_event === null
      || typeof (runtime_event as LangSmithRun).id !== 'string'
      || typeof (runtime_event as LangSmithRun).run_type !== 'string'
    ) {
      throw new AdapterTranslationError(
        `expected LangSmithRun object, got ${runtime_event === null ? 'null' : typeof runtime_event}`,
      );
    }

    const run = runtime_event;

    let resultStatus: string;
    if (run.error) {
      resultStatus = 'ERROR';
    } else {
      resultStatus = 'OK';
    }

    const kind = KNOWN_RUN_TYPES.has(run.run_type) ? run.run_type : 'unknown';

    const inputs = run.inputs ?? {};
    const payload: Record<string, unknown> = {
      kind,
      tool_name: `langsmith.${kind}.${run.name}`,
      tool_call_id: run.id,
      arguments_hash: _hashJson(inputs),
      result_status: resultStatus,
    };
    if (run.outputs != null) {
      payload['result_hash'] = _hashJson(run.outputs);
    }
    if (run.end_time != null) {
      const ms = run.end_time.getTime() - run.start_time.getTime();
      payload['latency_ms'] = Math.trunc(ms);
    }
    if (run.error) {
      payload['error_code'] = _truncate(run.error);
    }
    if (run.tags && run.tags.length > 0) {
      payload['tags'] = [...run.tags];
    }

    const sessionId = run.trace_id ?? run.id;

    return makeEventDraft({
      event_type: TOOL_CALL_EVENT,
      actor: `langsmith://${run.run_type}/${run.name}`,
      payload,
      subject_ref: run.end_user_id ? makeSubjectRef('opaque', run.end_user_id) : null,
      session_id: sessionId,
      reference_db_ref: run.parent_run_id ?? null,
    });
  }

  static fromDict(raw: Record<string, unknown>): LangSmithRun {
    const required = ['id', 'name', 'run_type', 'start_time'] as const;
    const missing = required.filter((k) => !(k in raw));
    if (missing.length > 0) {
      throw new AdapterTranslationError(
        `LangSmith run dict missing required fields: ${JSON.stringify(missing.sort())}`,
      );
    }

    const parseDt = (value: unknown): Date => {
      if (value instanceof Date) return value;
      if (typeof value === 'string') {
        const normalized = value.endsWith('Z')
          ? value.replace('Z', '+00:00')
          : value;
        const d = new Date(normalized);
        if (Number.isNaN(d.getTime())) {
          throw new AdapterTranslationError(`unparseable datetime ${JSON.stringify(value)}`);
        }
        return d;
      }
      throw new AdapterTranslationError(
        `datetime field has type ${typeof value}, expected Date or ISO string`,
      );
    };

    const metadata = (raw['metadata'] as Record<string, unknown> | undefined) ?? {};
    if (typeof metadata !== 'object' || Array.isArray(metadata)) {
      throw new AdapterTranslationError('metadata must be an object');
    }

    const tagsRaw = raw['tags'];
    if (tagsRaw !== undefined && tagsRaw !== null && !Array.isArray(tagsRaw)) {
      throw new AdapterTranslationError('tags must be an array');
    }
    const tags = Array.isArray(tagsRaw)
      ? tagsRaw.map((t) => String(t))
      : [];

    return {
      id: String(raw['id']),
      name: String(raw['name']),
      run_type: String(raw['run_type']),
      start_time: parseDt(raw['start_time']),
      end_time: raw['end_time'] ? parseDt(raw['end_time']) : null,
      inputs: (raw['inputs'] as Record<string, unknown>) ?? {},
      outputs: (raw['outputs'] as Record<string, unknown> | null) ?? null,
      error: (raw['error'] as string | null) ?? null,
      trace_id: (raw['trace_id'] as string | null) ?? null,
      parent_run_id: (raw['parent_run_id'] as string | null) ?? null,
      status: (raw['status'] as string | null) ?? null,
      tags,
      metadata,
      end_user_id: typeof metadata['user_id'] === 'string' ? metadata['user_id'] : null,
    };
  }
}
